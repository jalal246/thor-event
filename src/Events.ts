/* eslint-disable no-nested-ternary */
import Emitter from "./Emitter";
import Registry from "./Registry";

interface optsWithID {
  id: string;
  type?: never;
}

interface optsWithType<T extends string> {
  id?: never;
  type: T;
}

interface optsWithBoth<T extends string> {
  id: string;
  type: T;
}

class Events<
  EventTypes extends string = string,
  PayLoadInterface extends any = {},
  IssuerTypes extends string = string
> extends Registry<EventTypes, PayLoadInterface, IssuerTypes> {
  private emitter: Emitter<EventTypes>;

  constructor() {
    super();

    this.emitter = new Emitter<EventTypes>();
  }

  on(type: EventTypes, ...listeners: Function[]) {
    this.emitter.on(type, ...listeners);
  }

  off(type: EventTypes, listener: Function) {
    this.emitter.off(type, listener);
  }

  emit(opts: optsWithID | optsWithType<EventTypes> | optsWithBoth<EventTypes>) {
    const ids = opts.id
      ? [opts.id]
      : opts.type
      ? this.idsByType[opts.type]
      : [];

    ids.forEach((id) => {
      this.setPayload({ id, ...opts });

      const { type } = this.events[id];

      if (!type) return;

      this.emitter.emit(type, this.events[id]);
    });
  }
}

export default Events;
