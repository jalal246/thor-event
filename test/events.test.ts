import Events from "../src";
import type { EventInterface } from "../src";

type EventTypes = "type-1" | "type-2";
type IssuerTypes = "tester";
interface PayLoadInterface {
  name: string;
  address: string;
}

type EventI = EventInterface<EventTypes, PayLoadInterface, IssuerTypes>;

describe("Events", () => {
  let event: Events<EventTypes, PayLoadInterface, IssuerTypes>;

  const mockCallback1 = jest.fn(() => "function-1 is called");
  const mockCallback2 = jest.fn(() => "function-2 is called");
  const mockCallback3 = jest.fn(() => "function-3 is called");

  const evt1: EventI = {
    id: "id-1",
    type: "type-1",
    issuer: "tester",
    payload: {
      name: "name-1",
      address: "address-1",
    },
  };

  const evt2: EventI = {
    id: "id-2",
    type: "type-1",
    issuer: "tester",
    payload: {
      name: "name-2",
      address: "address-2",
    },
  };

  const evt3: EventI = {
    id: "id-3",
    type: "type-2",
    issuer: "tester",
    payload: {
      name: "name-3",
      address: "address-3",
    },
  };

  beforeAll(() => {
    jest.useFakeTimers("modern");
    jest.setSystemTime(1648764000000);

    event = new Events();
  });

  it("should be defined", () => {
    expect(event).toMatchSnapshot();
  });

  it("should initialize event with two different types", () => {
    event.init(evt1);
    event.init(evt2);
    event.init(evt3);

    expect(event).toMatchSnapshot();
  });

  it("should update event-3 payload", () => {
    event.setPayload({
      id: evt3.id,
      payload: {
        name: "name-updated-3",
        address: "address-updated-3",
      },
    });

    expect(event).toMatchSnapshot();
  });

  it("should update event-2 id", () => {
    event.updateID({
      id: evt3.id,
      newID: "id-3-updated",
    });

    expect(event.idsByType).toMatchObject({
      "type-1": ["id-1", "id-2"],
      "type-2": ["id-3-updated"],
    });
  });

  it("should return all registered events in the same type", () => {
    expect(event.getEventsByType("type-1")).toMatchObject(["id-1", "id-2"]);
    expect(event.getEventsByType("type-2")).toMatchObject(["id-3-updated"]);
  });

  it("should clear from registry", () => {
    event.clear("id-3-updated");
    expect(event.getEventsByType("type-2")).toMatchObject([]);
    expect(event.idsByType).toMatchObject({
      "type-1": ["id-1", "id-2"],
    });
  });

  it("should add functions emitter", () => {
    event.on("type-1", mockCallback1, mockCallback2, mockCallback3);
    expect(event).toMatchSnapshot();
  });

  it("should emit all registered function by type", () => {
    event.emit({
      type: "type-1",
    });

    expect(mockCallback1).toHaveBeenCalledTimes(2);
    expect(mockCallback2).toHaveBeenCalledTimes(2);
    expect(mockCallback3).toHaveBeenCalledTimes(2);
  });

  it("should emit all registered function by id", () => {
    event.emit({
      id: evt1.id,
    });

    expect(mockCallback1).toHaveBeenCalledTimes(1);
    expect(mockCallback2).toHaveBeenCalledTimes(1);
    expect(mockCallback3).toHaveBeenCalledTimes(1);
  });
});
