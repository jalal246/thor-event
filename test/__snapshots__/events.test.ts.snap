// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Events should add functions emitter 1`] = `
Events {
  "emitter": Emitter {
    "listeners": Object {
      "type-1": Array [
        [MockFunction],
        [MockFunction],
        [MockFunction],
      ],
    },
  },
  "events": Object {
    "id-1": Object {
      "createdAt": 1648764000000,
      "issuer": "tester",
      "payload": Object {
        "address": "address-1",
        "name": "name-1",
      },
      "type": "type-1",
    },
    "id-2": Object {
      "createdAt": 1648764000000,
      "issuer": "tester",
      "payload": Object {
        "address": "address-2",
        "name": "name-2",
      },
      "type": "type-1",
    },
  },
  "idsByType": Object {
    "type-1": Array [
      "id-1",
      "id-2",
    ],
  },
}
`;

exports[`Events should be defined 1`] = `
Events {
  "emitter": Emitter {
    "listeners": Object {},
  },
  "events": Object {},
  "idsByType": Object {},
}
`;

exports[`Events should initialize event with two different types 1`] = `
Events {
  "emitter": Emitter {
    "listeners": Object {},
  },
  "events": Object {
    "id-1": Object {
      "createdAt": 1648764000000,
      "issuer": "tester",
      "payload": Object {
        "address": "address-1",
        "name": "name-1",
      },
      "type": "type-1",
    },
    "id-2": Object {
      "createdAt": 1648764000000,
      "issuer": "tester",
      "payload": Object {
        "address": "address-2",
        "name": "name-2",
      },
      "type": "type-1",
    },
    "id-3": Object {
      "createdAt": 1648764000000,
      "issuer": "tester",
      "payload": Object {
        "address": "address-3",
        "name": "name-3",
      },
      "type": "type-2",
    },
  },
  "idsByType": Object {
    "type-1": Array [
      "id-1",
      "id-2",
    ],
    "type-2": Array [
      "id-3",
    ],
  },
}
`;

exports[`Events should update event-3 payload 1`] = `
Events {
  "emitter": Emitter {
    "listeners": Object {},
  },
  "events": Object {
    "id-1": Object {
      "createdAt": 1648764000000,
      "issuer": "tester",
      "payload": Object {
        "address": "address-1",
        "name": "name-1",
      },
      "type": "type-1",
    },
    "id-2": Object {
      "createdAt": 1648764000000,
      "issuer": "tester",
      "payload": Object {
        "address": "address-2",
        "name": "name-2",
      },
      "type": "type-1",
    },
    "id-3": Object {
      "createdAt": 1648764000000,
      "issuer": "tester",
      "payload": Object {
        "address": "address-3",
        "name": "name-3",
      },
      "type": "type-2",
    },
  },
  "idsByType": Object {
    "type-1": Array [
      "id-1",
      "id-2",
    ],
    "type-2": Array [
      "id-3",
    ],
  },
}
`;
